File: LeetCode/Python/move_zeroes.py
"""
Problem:
Given an array nums, write a function to move all 0's to the end of it while
maintaining the relative order of the non-zero elements.

Example:
Input: [0, 1, 0, 3, 12]
Output: [1, 3, 12, 0, 0]

Note:
- You must do this in-place without making a copy of the array.
- Minimize the total number of operations.
"""


def moveZeroes(nums: [int]) -> None:
    """
    Do not return anything, modify nums in-place instead.
    """
    nonZeroIdx = 0
    for i in range(len(nums)):
        # if the current element is not 0
        if nums[i] != 0:
            # put the non-zero elements at the beginning of the list
            nums[nonZeroIdx] = nums[i]
            nonZeroIdx += 1
    # fill in the remainder of the list with 0's
    for i in range(nonZeroIdx, len(nums)):
        nums[i] = 0

File: LeetCode/Java/move_zeroes.java
/**
 * Given an array nums, write a function to move all 0's to the end of it while
 * maintaining the relative order of the non-zero elements.
 * 
 * @param nums The array to modify
 * @return Do not return anything, modify nums in-place instead
 */

class Solution {
  public void moveZeroes(int[] nums) {
    int nonZeroIdx = 0;
    for (int i = 0; i < nums.length; i++) {
      // if the current element is not 0
      if (nums[i] != 0) {
        // append non-zero elements in front of the previous non-zero elements
        nums[nonZeroIdx++] = nums[i];
      }
    }
    // fill in the remainder of the array with 0's
    for (int i = nonZeroIdx; i < nums.length; i++) {
      nums[i] = 0;
    }
  }
}